File: src/Schema.ts
export class SchemaValidationError {
    expected : string[];
    received : string;
    property ?: string;

    constructor ( expected : string[] | string, received : string, property : string = null ) {
        this.property = property;
        
        if ( expected instanceof Array ) {
            this.expected = expected;
        } else {
            this.expected = [ expected ];
        }

        this.received = received;
    }

    prefix ( property : string ) : SchemaValidationError {
        if ( this.property == null ) {
            return new SchemaValidationError( this.expected, this.received, property );
        }

        return new SchemaValidationError( this.expected, this.received, property + '.' + this.property );
    }

    get message () {
        const expectations = `Expected ${ this.expected.join( ', ' ) }, got ${ this.received } instead.`;

        if ( this.property !== null ) {
            return `${ this.property }: ${ expectations }`
        } else {
            return expectations;
        }
    }


    public static toString ( errors ?: SchemaValidationResult ) {
        if ( !errors ) {
            return  null;
        }

        if ( errors instanceof Array ) {
            return errors.map( error => error.message ).join( '\n' );
        } else {
            return errors.message;
        }
    }
}

export type SchemaValidationResult = null | SchemaValidationError | SchemaValidationError[];

export abstract class TypeSchema {
    static normalize ( schema : any ) : TypeSchema {
        if ( schema instanceof TypeSchema ) {
            return schema;
        } else if ( schema instanceof Array ) {
            if ( schema.length == 0 ) {
                return ArrayTypeSchema.normalize( new AnyTypeSchema() );
            } else if ( schema.length == 1 ) {
                return ArrayTypeSchema.normalize( schema );
            } else {
                return TupleTypeSchema.normalize( schema );
            }
        } else if ( schema === String ) {
            return new StringTypeSchema();
        } else if ( schema === Number ) {
            return new NumberTypeSchema();
        } else if ( schema === Boolean ) {
            return new BooleanTypeSchema();
        } else if ( typeof schema === 'object' ) {
            return ObjectTypeSchema.normalize( schema );
        } else {
            return new ConstantTypeSchema( schema );
        }
    }

    abstract validate ( data : any ) : SchemaValidationResult;

    abstract run ( data : any ) : any;
}


export class ConstantTypeSchema extends TypeSchema {
    constant : any = null;

    constructor ( constant : any ) {
        super();

        this.constant = constant;
    }

    validate ( data : any ) : SchemaValidationResult {
        if ( data == this.constant ) {
            return null;
        }

        return new SchemaValidationError( `"${ this.constant }"`, `"${ data }"` );
    }

    run ( data : any ) {
        return this.constant;
    }
}

export class OptionalTypeSchema extends TypeSchema {
    subSchema : TypeSchema;

    defaultValue : any = null;

    constructor ( subSchema : any, defaultValue : any = null ) {
        super();

        this.subSchema = TypeSchema.normalize( subSchema );

        this.defaultValue = defaultValue;
    }

    validate ( data : any ) : SchemaValidationResult {
        if ( data === null || data === void 0 ) {
            return null;
        }

        return this.subSchema.validate( data );
    }

    run ( data : any ) {
        if ( data === null || data === void 0 ) {
            data = this.defaultValue;
        }

        return this.subSchema.run( data );
    }
}

export class AnyTypeSchema extends TypeSchema {
    validate () {
        return null;
    }

    run ( data : any ) {
        return data;
    }
}

export class UnionTypeSchema extends TypeSchema {
    typeSchemas: TypeSchema[];
    
    constructor ( ...typeSchemas: any[] ) {
        super();

        this.typeSchemas = typeSchemas.map( type => TypeSchema.normalize( type ) );
    }

    validate ( data : any ) : SchemaValidationResult {
        const errors: SchemaValidationError[] = [];

        for ( const schema of this.typeSchemas ) {
            const schemaErrors = schema.validate( data );

            if ( schemaErrors === null ) {
                return null;
            }

            if ( schemaErrors instanceof Array ) {
                errors.push( ...schemaErrors );
            } else {
                errors.push( schemaErrors );
            }
        }

        if ( errors.length === 0 ) {
            return null;
        }

        return errors;
    }

    run ( data : any ) {
        for ( const schema of this.typeSchemas ) {
            const schemaErrors = schema.validate( data );

            if ( schemaErrors === null ) {
                return schema.run( schema );
            }
        }

        return data;
    }
}

export class IntersectionTypeSchema extends TypeSchema {
    typeSchemas: TypeSchema[];
    
    constructor ( ...typeSchemas: any[] ) {
        super();

        this.typeSchemas = typeSchemas.map( type => TypeSchema.normalize( type ) );
    }

    validate ( data : any ) : SchemaValidationResult {
        const errors: SchemaValidationError[] = [];

        for ( const schema of this.typeSchemas ) {
            const schemaErrors = schema.validate( data );

            if ( schemaErrors instanceof Array ) {
                errors.push( ...schemaErrors );
            } else if ( schemaErrors != null ) {
                errors.push( schemaErrors );
            }
        }

        if ( errors.length === 0 ) {
            return null;
        }

        return errors;
    }

    run ( data : any ) {
        for ( const schema of this.typeSchemas ) {
            data = schema.run( data );
        }

        return data;
    }
}

export class StringTypeSchema extends TypeSchema {
    validate ( data : any ) : SchemaValidationResult {
        if ( typeof data === 'string' ) {
            return null;
        }

        return new SchemaValidationError( 'String', typeof data );
    }

    run ( data : any ) {
        return data;
    }
}

export class NumberTypeSchema extends TypeSchema {
    validate ( data : any ) {
        if ( typeof data === 'number' ) {
            return null;
        }

        return new SchemaValidationError( 'Number', typeof data );
    }

    run ( data : any ) {
        return data;
    }
}


export class BooleanTypeSchema extends TypeSchema {
    validate ( data : any ) {
        if ( typeof data === 'boolean' ) {
            return null;
        }

        return new SchemaValidationError( 'Boolean', typeof data );
    }

    run ( data : any ) {
        return data;
    }
}

export class TupleTypeSchema extends TypeSchema {
    static normalize ( schema : any ) : TupleTypeSchema {
        return new TupleTypeSchema( schema );
    }

    subSchema : TypeSchema[];

    constructor ( schema : any[] ) {
        super();

        this.subSchema = [];

        for ( let type of schema ) {
            this.subSchema.push( TypeSchema.normalize( type ) );
        }
    }

    validate ( data : any ) : SchemaValidationResult {
        if ( data instanceof Array ) {
            const errors = data.map( ( item, index ) => {
                    if ( this.subSchema.length <= index ) {
                        return new SchemaValidationError( 'Undefined', typeof item );
                    }

                    const errors = this.subSchema[ index ].validate( item );

                    if ( errors instanceof Array ) {
                        return errors.map( err => err.prefix( index.toString() ) );
                    } else if ( errors !== null ) {
                        return errors.prefix( index.toString() );
                    }
                } ).filter( error => error != null )
                .reduce( ( arr, errors ) => {
                    if ( errors instanceof Array ) {
                        arr.push( ...errors );
                    } else {
                        arr.push( errors )
                    }

                    return arr;
                }, [] as any[] );

                if ( errors.length === 0 ) {
                    return null;
                }

                return errors;
        }

        return new SchemaValidationError( 'Array', typeof data );
    }

    run ( data : any ) {
        if ( data instanceof Array ) {
            return data.map( ( entry, index ) => {
                if ( index < this.subSchema.length ) {
                    return this.subSchema[ index ].run( entry )
                }

                return entry;
            } );
        }
        
        return data;
    }
}

export class ArrayTypeSchema extends TypeSchema {
    static normalize ( schema : any ) : TypeSchema {
        return new ArrayTypeSchema( TypeSchema.normalize( schema[ 0 ] ) );
    }

    subSchema : TypeSchema;

    constructor ( subSchema : any ) {
        super();

        this.subSchema = TypeSchema.normalize( subSchema );
    }

    validate ( data : any ) : SchemaValidationResult {
        if ( data instanceof Array ) {
            const errors = data.map( ( item, index ) => {
                    const errors = this.subSchema.validate( item );

                    if ( errors instanceof Array ) {
                        return errors.map( err => err.prefix( index.toString() ) );
                    } else if ( errors !== null ) {
                        return errors.prefix( index.toString() );
                    }
                } )
                .filter( error => error != null )
                .reduce( ( arr, errors ) => {
                    if ( errors instanceof Array ) {
                        arr.push( ...errors );
                    } else {
                        arr.push( errors )
                    }

                    return arr;
                }, [] as any[] );

                if ( errors.length === 0 ) {
                    return null;
                }

                return errors;
        }

        return new SchemaValidationError( 'Array', typeof data );
    }

    run ( data : any ) {
        if ( data instanceof Array ) {
            return data.map( entry => this.subSchema.run( entry ) );
        }
        
        return data;
    }
}

export class ObjectTypeSchema extends TypeSchema {
    static normalize ( schema : any ) : TypeSchema {
        return new ObjectTypeSchema( schema );
    }

    subSchema : { [ key : string] : TypeSchema };

    // If true, keys that are not defined in the schema are not allowed
    strict : boolean;

    constructor ( subSchema : any, strict : boolean = false ) {
        super();

        this.subSchema = {};

        for ( let key of Object.keys( subSchema ) ) {
            this.subSchema[ key ] = TypeSchema.normalize( subSchema[ key ] );
        }

        this.strict = strict;
    }

    validate ( data : any ) : SchemaValidationResult {
        if ( data && typeof data === 'object' ) {
            const errors : SchemaValidationError[] = [];

            const requiredKeys = new Set( Object.keys( this.subSchema ) );

            for ( let key of Object.keys( data ) ) {
                if ( !( key in this.subSchema ) && this.strict ) {
                    errors.push( new SchemaValidationError( 'Undefined', typeof data[ key ], key ) )
                } else if ( key in this.subSchema ) {
                    requiredKeys.delete( key );

                    const keyErrors = this.subSchema[ key ].validate( data[ key ] );

                    if ( keyErrors instanceof Array ) {
                        errors.push( ...keyErrors.map( error => error.prefix( key ) ) );
                    } else if ( keyErrors !== null ) {
                        errors.push( keyErrors.prefix( key ) );
                    }
                }
            }

            for ( let key of requiredKeys ) {
                const keyErrors = this.subSchema[ key ].validate( void 0 );

                if ( keyErrors instanceof Array ) {
                    errors.push( ...keyErrors.map( error => error.prefix( key ) ) );
                } else if ( keyErrors !== null ) {
                    errors.push( keyErrors.prefix( key ) );
                }
            }

            if ( errors.length === 0 ) {
                return null;
            }

            return errors;
        }

        return new SchemaValidationError( 'Object', typeof data );
    }

    run ( data : any ) {
        const requiredKeys = new Set( Object.keys( this.subSchema ) );

        const result : any = {};

        for ( let key of Object.keys( data ) ) {
            if ( key in this.subSchema ) {
                requiredKeys.delete( key );

                result[ key ] = this.subSchema[ key ].run( data[ key ] );
            } else {
                result[ key ] = data[ key ];
            }
        }

        for ( let key of requiredKeys ) {
            result[ key ] = this.subSchema[ key ].run( data[ key ] );
        }

        return result;
    }
}

// Functional DSL
export function any () {
    return new AnyTypeSchema();
}

export function array ( subSchema : any ) {
    return new ArrayTypeSchema( subSchema );
}

export function object ( subSchema : any = {}, strict : boolean = false ) {
    return new ObjectTypeSchema( subSchema, strict );
}

export function union ( ...typeSchemas : any[] ) {
    return new UnionTypeSchema( ...typeSchemas );
}

export function intersection ( ...typeSchemas : any[] ) {
    return new IntersectionTypeSchema( ...typeSchemas );
}

export function tuple ( subSchema : any ) {
    return new TupleTypeSchema( subSchema );
}

export function constant ( constant : any ) {
    return new ConstantTypeSchema( constant );
}

export function number () {
    return new NumberTypeSchema();
}

export function boolean () {
    return new BooleanTypeSchema();
}

export function optional ( subSchema : any, defaultValue : any = null ) {
    return new OptionalTypeSchema( subSchema, defaultValue );
}

export function string () {
    return new StringTypeSchema();
}

File: pkg/schema/schema.go
package schema

import (
	"fmt"
	"reflect"
	"strconv"
	"strings"
)

type ValidationError struct {
	Expected []string
	Received string
	Property string
	Errors   []*ValidationError
}

func (e *ValidationError) Error() string {
	if len(e.Errors) > 0 {
		var msgs []string
		for _, err := range e.Errors {
			msgs = append(msgs, err.Error())
		}
		return strings.Join(msgs, "\n")
	}
	expectations := fmt.Sprintf("Expected %s, got %s instead.",
		strings.Join(e.Expected, ", "), e.Received)
	if e.Property != "" {
		return fmt.Sprintf("%s: %s", e.Property, expectations)
	}
	return expectations
}

func (e *ValidationError) Prefix(property string) *ValidationError {
	if len(e.Errors) > 0 {
		prefixed := make([]*ValidationError, len(e.Errors))
		for i, err := range e.Errors {
			prefixed[i] = err.Prefix(property)
		}
		return &ValidationError{Errors: prefixed}
	}
	prop := property
	if e.Property != "" {
		prop = property + "." + e.Property
	}
	return &ValidationError{
		Expected: e.Expected,
		Received: e.Received,
		Property: prop,
	}
}

func ToString(err *ValidationError) string {
	if err == nil {
		return ""
	}
	return err.Error()
}

type Schema interface {
	Validate(data interface{}) *ValidationError
}

func typeName(data interface{}) string {
	if data == nil {
		return "nil"
	}
	switch data.(type) {
	case string:
		return "string"
	case bool:
		return "boolean"
	case int, int8, int16, int32, int64,
		uint, uint8, uint16, uint32, uint64,
		float32, float64:
		return "number"
	default:
		r := reflect.ValueOf(data)
		switch r.Kind() {
		case reflect.Slice, reflect.Array:
			return "object"
		case reflect.Map:
			return "object"
		case reflect.Struct:
			return "object"
		default:
			return r.Kind().String()
		}
	}
}

func formatValue(v interface{}) string {
	if v == nil {
		return "nil"
	}
	switch val := v.(type) {
	case string:
		return fmt.Sprintf("\"%s\"", val)
	default:
		return fmt.Sprintf("%v", val)
	}
}

func toSlice(data interface{}) ([]interface{}, bool) {
	if data == nil {
		return nil, false
	}
	v := reflect.ValueOf(data)
	if v.Kind() != reflect.Slice && v.Kind() != reflect.Array {
		return nil, false
	}
	result := make([]interface{}, v.Len())
	for i := 0; i < v.Len(); i++ {
		result[i] = v.Index(i).Interface()
	}
	return result, true
}

func toMap(data interface{}) (map[string]interface{}, bool) {
	if data == nil {
		return nil, false
	}
	v := reflect.ValueOf(data)
	if v.Kind() != reflect.Map {
		return nil, false
	}
	result := make(map[string]interface{})
	iter := v.MapRange()
	for iter.Next() {
		key := iter.Key()
		if key.Kind() == reflect.String {
			result[key.String()] = iter.Value().Interface()
		}
	}
	return result, true
}

func prefixErrors(err *ValidationError, prefix string) []*ValidationError {
	if err == nil {
		return nil
	}
	if len(err.Errors) > 0 {
		var result []*ValidationError
		for _, e := range err.Errors {
			result = append(result, prefixErrors(e, prefix)...)
		}
		return result
	}
	return []*ValidationError{err.Prefix(prefix)}
}

func collectLeafErrors(err *ValidationError) []*ValidationError {
	if err == nil {
		return nil
	}
	if len(err.Errors) > 0 {
		var result []*ValidationError
		for _, e := range err.Errors {
			result = append(result, collectLeafErrors(e)...)
		}
		return result
	}
	return []*ValidationError{err}
}

// StringSchema validates that data is a string.
type StringSchema struct{}

func (s *StringSchema) Validate(data interface{}) *ValidationError {
	if _, ok := data.(string); ok {
		return nil
	}
	return &ValidationError{Expected: []string{"String"}, Received: typeName(data)}
}

func String() *StringSchema { return &StringSchema{} }

// NumberSchema validates that data is a numeric type.
type NumberSchema struct{}

func (s *NumberSchema) Validate(data interface{}) *ValidationError {
	switch data.(type) {
	case int, int8, int16, int32, int64,
		uint, uint8, uint16, uint32, uint64,
		float32, float64:
		return nil
	}
	return &ValidationError{Expected: []string{"Number"}, Received: typeName(data)}
}

func Number() *NumberSchema { return &NumberSchema{} }

// BooleanSchema validates that data is a boolean.
type BooleanSchema struct{}

func (s *BooleanSchema) Validate(data interface{}) *ValidationError {
	if _, ok := data.(bool); ok {
		return nil
	}
	return &ValidationError{Expected: []string{"Boolean"}, Received: typeName(data)}
}

func Boolean() *BooleanSchema { return &BooleanSchema{} }

// AnySchema accepts any data without validation.
type AnySchema struct{}

func (s *AnySchema) Validate(data interface{}) *ValidationError {
	return nil
}

func Any() *AnySchema { return &AnySchema{} }

// OptionalSchema validates data against a sub-schema if present, or accepts nil.
type OptionalSchema struct {
	SubSchema    Schema
	DefaultValue interface{}
}

func (s *OptionalSchema) Validate(data interface{}) *ValidationError {
	if data == nil {
		return nil
	}
	return s.SubSchema.Validate(data)
}

func Optional(sub Schema, defaultVal ...interface{}) *OptionalSchema {
	s := &OptionalSchema{SubSchema: sub}
	if len(defaultVal) > 0 {
		s.DefaultValue = defaultVal[0]
	}
	return s
}

// ConstantSchema validates that data equals a specific constant value.
type ConstantSchema struct {
	Value interface{}
}

func (s *ConstantSchema) Validate(data interface{}) *ValidationError {
	if reflect.DeepEqual(data, s.Value) {
		return nil
	}
	return &ValidationError{
		Expected: []string{formatValue(s.Value)},
		Received: formatValue(data),
	}
}

func Constant(val interface{}) *ConstantSchema { return &ConstantSchema{Value: val} }

// TupleSchema validates an array where each position matches a specific schema.
type TupleSchema struct {
	Schemas []Schema
}

func (s *TupleSchema) Validate(data interface{}) *ValidationError {
	arr, ok := toSlice(data)
	if !ok {
		return &ValidationError{Expected: []string{"Array"}, Received: typeName(data)}
	}
	var errors []*ValidationError
	schemaLen := len(s.Schemas)
	for i := 0; i < schemaLen; i++ {
		if i >= len(arr) {
			if err := s.Schemas[i].Validate(nil); err != nil {
				errors = append(errors, prefixErrors(err, strconv.Itoa(i))...)
			}
		} else if err := s.Schemas[i].Validate(arr[i]); err != nil {
			errors = append(errors, prefixErrors(err, strconv.Itoa(i))...)
		}
	}
	for i := schemaLen; i < len(arr); i++ {
		errors = append(errors, &ValidationError{
			Expected: []string{"Undefined"},
			Received: typeName(arr[i]),
			Property: strconv.Itoa(i),
		})
	}
	if len(errors) == 0 {
		return nil
	}
	return &ValidationError{Errors: errors}
}

func Tuple(schemas ...Schema) *TupleSchema {
	return &TupleSchema{Schemas: schemas}
}

// ArraySchema validates that data is an array with all items matching a sub-schema.
type ArraySchema struct {
	SubSchema Schema
}

func (s *ArraySchema) Validate(data interface{}) *ValidationError {
	arr, ok := toSlice(data)
	if !ok {
		return &ValidationError{Expected: []string{"Array"}, Received: typeName(data)}
	}
	var errors []*ValidationError
	for i, item := range arr {
		if err := s.SubSchema.Validate(item); err != nil {
			errors = append(errors, prefixErrors(err, strconv.Itoa(i))...)
		}
	}
	if len(errors) == 0 {
		return nil
	}
	return &ValidationError{Errors: errors}
}

func Array(sub Schema) *ArraySchema { return &ArraySchema{SubSchema: sub} }

// ObjectSchema validates an object with typed properties.
type ObjectSchema struct {
	Properties map[string]Schema
	Strict     bool
}

func (s *ObjectSchema) Validate(data interface{}) *ValidationError {
	m, ok := toMap(data)
	if !ok {
		return &ValidationError{Expected: []string{"Object"}, Received: typeName(data)}
	}
	var errors []*ValidationError

	requiredKeys := make(map[string]bool)
	for k := range s.Properties {
		requiredKeys[k] = true
	}

	for k, v := range m {
		if propSchema, exists := s.Properties[k]; exists {
			delete(requiredKeys, k)
			if err := propSchema.Validate(v); err != nil {
				errors = append(errors, prefixErrors(err, k)...)
			}
		} else if s.Strict {
			errors = append(errors, &ValidationError{
				Expected: []string{"Undefined"},
				Received: typeName(v),
				Property: k,
			})
		}
	}

	for k := range requiredKeys {
		if err := s.Properties[k].Validate(nil); err != nil {
			errors = append(errors, prefixErrors(err, k)...)
		}
	}

	if len(errors) == 0 {
		return nil
	}
	return &ValidationError{Errors: errors}
}

func Object(props map[string]Schema, strict ...bool) *ObjectSchema {
	s := &ObjectSchema{Properties: props}
	if len(strict) > 0 {
		s.Strict = strict[0]
	}
	return s
}

// UnionSchema validates that data matches at least one sub-schema.
type UnionSchema struct {
	Schemas []Schema
}

func (s *UnionSchema) Validate(data interface{}) *ValidationError {
	var allErrors []*ValidationError
	for _, schema := range s.Schemas {
		if err := schema.Validate(data); err == nil {
			return nil
		} else {
			allErrors = append(allErrors, collectLeafErrors(err)...)
		}
	}
	if len(allErrors) == 0 {
		return &ValidationError{Expected: []string{"any of union"}, Received: typeName(data)}
	}
	return &ValidationError{Errors: allErrors}
}

func Union(schemas ...Schema) *UnionSchema {
	return &UnionSchema{Schemas: schemas}
}

// IntersectionSchema validates that data matches all sub-schemas.
type IntersectionSchema struct {
	Schemas []Schema
}

func (s *IntersectionSchema) Validate(data interface{}) *ValidationError {
	var allErrors []*ValidationError
	for _, schema := range s.Schemas {
		if err := schema.Validate(data); err != nil {
			allErrors = append(allErrors, collectLeafErrors(err)...)
		}
	}
	if len(allErrors) == 0 {
		return nil
	}
	return &ValidationError{Errors: allErrors}
}

func Intersection(schemas ...Schema) *IntersectionSchema {
	return &IntersectionSchema{Schemas: schemas}
}